### PR TITLE
EOS-13370: Handle special stateid for setattr

### PR DESCRIPTION
**Pull request by sachinpunadikar**
# CORTXFS_FSAL Change Summary

## Problem Statement

_[EOS-13370](https://jts.seagate.com/browse/EOS-13370):_
_Fix the Ganesha crash for NULL stateid.._

## Problem Description

_In certain cases NFS client may send a special stateid (which is NULL) indicating that the file is not opened and ahs no state associated. Need to add code for handling such situations._

## Solution Overview

_Added code to handle special stateid for setattr operation._
In certain cases NFS client passes special stateid (which is NULL).
Need to consider such scenarios (meaning the file has not been opened) and
work on the requested operation. This patch addresses such scenario for
truncate operation.

## Unit Test Cases
_Verified that nfstest_io runs successfully without crashing Ganesha_
_cthon test is passing Basic test cases_

## Checklist
- [x] **Compilation:** _This patch does not break compilation_
- [x] **Merge conflicts:** _This patch has been squashed and re-based, it can be merged using fast-forward merge_
- [x] **Code review:** _All discussions have been resolved_
- [x] **Sanity Testing:** _As mentioned related test cases are passing_

### CODE REVIEW

File: src/FSAL/FSAL_CORTXFS/handle.c [hunk 1]
```diff
@@ -2833,37 +2864,52 @@ static fsal_status_t kvsfs_ftruncate(struct fsal_obj_handle *obj_hdl,
...
-	 */
-	result = kvsfs_find_fd(state, bypass, FSAL_O_WRITE, &fd);
-	if (FSAL_IS_ERROR(result)) {
-		goto out;
+	obj = container_of(obj_hdl, struct kvsfs_fsal_obj_handle, obj_handle);
+
+	if (state == NULL) {
+		/* special stateid handling required */
+		fd = gsh_calloc(1, sizeof(struct kvsfs_file_state));
+		result = kvsfs_get_special_fd(obj, FSAL_O_WRITE, fd);
```
Comment (pratyush-seagate) [inline src/FSAL/FSAL_CORTXFS/handle.c:2885]: Just a doubt, is this fix needed to other callers of kvsfs_find_fd() where arg state can be NULL, i.e. kvsfs_read2, kvsfs_write2.

**Review by Anilsindhi — APPROVED** [review]
Looks good to me.

**Review by s-arya — APPROVED** [review]
Looks good to me.

File: src/FSAL/FSAL_CORTXFS/handle.c (hunk 1 above)
Comment (sachinpunadikar) [reply src/FSAL/FSAL_CORTXFS/handle.c:2885]: Hi Pratyush, your understanding is correct. This fix would be needed for those functions as well.
Just holding it till we have clean test results for this change. Once everything looks good, we will have this fix for read & write paths as well.

Comment (pratyush-seagate) [reply src/FSAL/FSAL_CORTXFS/handle.c:2885]: Thanks Sachin for explaining this.

**Review by pratyush-seagate — APPROVED** [review]
LGTM, +1.

**Pull request merged**